File: algorithms/solutions/week26/duythinht/find_the_difference.c
/*
 * Problem: https://leetcode.com/problems/find-the-difference/description/
 */
#include <stdlib.h>

#define ALPHABET_SIZE 26

char findTheDifference(char* s, char* t) {
    int *char_count = (int*)calloc(ALPHABET_SIZE+'a', sizeof(int));
    while (*s) {
        char_count[*s++]--;
        char_count[*t++]++;
    }
    char_count[*t]++;
    for (char i='a'; i < ALPHABET_SIZE+'a'; i++) {
        if (char_count[i]) return i;
    }
    return 0;
}

File: algorithms/solutions/week26/manhdv/find_the_difference.rb
# https://leetcode.com/problems/find-the-difference/description/

def characters_hash(str)
  hash = Hash.new(0)

  0.upto(str.size - 1) do |idx|
    hash[str[idx]] += 1
  end

  hash
end

# @param {String} s
# @param {String} t
# @return {Character}
def find_the_difference(s, t)
  s_hash = characters_hash(s)
  t_hash = characters_hash(t)

  t_hash.reject { |k, v| v == s_hash[k] }.keys.first
end
